scoring/src/main/scala/ai/h2o/sparkling/ml/metrics/H2OBinomialMetrics.scala

```diff
@@ -0,0 +1,159 @@
+
+  def this() = this(Identifiable.randomUID("H2OBinomialMetrics"))
+}
+
+object H2OBinomialMetrics extends MetricCalculation {
+
+  /**
+    * The method calculates binomial metrics on a provided data frame with predictions and actual values.
+    *
+    * @param dataFrame A data frame with predictions and actual values
+    * @param domain Array of classes representing negative and positive response. Negative class must at position 0 and
+    *               positive at 1.
+    * @param predictionCol      The name of prediction column. The prediction column must have the same type as
+    *                           a detailed_prediction column coming from the transform method of H2OMOJOModel descendant.
+    *                           Or the type must be FloatType/DoubleType where values represent probabilities of
+    *                           the positive response.
+    * @param labelCol           The name of label column that contains actual values.
+    * @param weightColOption    The name of a weight column.
+    * @param offsetColOption    The name of a offset column.
+    * @return Calculated binomial metrics
+    */
+  def calculate(
+      dataFrame: DataFrame,
+      domain: Array[String],
+      predictionCol: String = "detailed_prediction",
+      labelCol: String = "label",
+      weightColOption: Option[String] = None,
+      offsetColOption: Option[String] = None): H2OBinomialMetrics = {
+    validateDataFrameForMetricCalculation(dataFrame, domain, predictionCol, labelCol, offsetColOption, weightColOption)
+    val getMetricBuilder = () => new MetricBuilderBinomial(domain)
+    val castedLabelDF = dataFrame.withColumn(labelCol, col(labelCol) cast StringType)
+
+    val gson =
+      getMetricGson(getMetricBuilder, castedLabelDF, predictionCol, labelCol, offsetColOption, weightColOption, domain)
+    val result = new H2OBinomialMetrics()
+    result.setMetrics(gson, "H2OBinomialMetrics.calculate")
+    result
+  }
+
+  // The method serves for call from Python API
+  def calculateInternal(
+      dataFrame: DataFrame,
+      domain: java.util.ArrayList[String],
+      predictionCol: String,
+      labelCol: String,
+      weightCol: String,
+      offsetCol: String): H2OBinomialMetrics = {
+    calculate(
+      dataFrame,
+      domain.toArray[String](new Array[String](0)),
+      predictionCol,
+      labelCol,
+      Option(weightCol),
+      Option(offsetCol))
+  }
+
+  // The method serves for call from R API
+  def calculateInternal(
+      dataFrame: DataFrame,
+      domain: Array[String],
+      predictionCol: String,
+      labelCol: String,
+      weightCol: String,
+      offsetCol: String): H2OBinomialMetrics = {
+    calculate(dataFrame, domain, predictionCol, labelCol, Option(weightCol), Option(offsetCol))
+  }
+
+  override protected def getPredictionValues(dataType: DataType, domain: Array[String], row: Row): Array[Double] = {
+    dataType match {
+      case StructType(fields)
```

Comment:
nit: matching like `StructType(Array(firstField, secondField))` would also ensure the field counts and avoid calling by index later on

Reply:
IMHO, the original approach is more readable. The pattern for the suggested approach would be:
```
StructType(Array(StructField(_, StringType, _, _), StructField(_, StructType(Array(StructField(_, DoubleType, _, _), StructField(_, DoubleType, _, _))), _, _)))
```

Reply:
The following "long" predicate appears several times in code. Would it be possible to define an auxiliary function that describes the condition and use it, e.g.
`def isSomeSpecialType(fields: Array[StructField]): Boolean =
  fields(0).dataType == StringType && fields(1).dataType.isInstanceOf[StructType] &&
  fields(1).dataType.asInstanceOf[StructType].fields.forall(_.dataType == DoubleType) &&
  fields(1).dataType.asInstanceOf[StructType].fields.length == 2`